hoomd/md/EvaluatorConstraintManifold.h

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2009-2019 The Regents of the University of Michigan
+// This file is part of the HOOMD-blue project, released under the BSD 3-Clause License.
+
+
+// Maintainer: joaander
+
+#ifndef __EVALUATOR_CONSTRAINT_Manifold_H__
+#define __EVALUATOR_CONSTRAINT_Manifold_H__
+
+#include "hoomd/HOOMDMath.h"
+#include "hoomd/ManifoldSurfaces.h"
+using namespace std;
+
+/*! \file EvaluatorConstraintManifold.h
+    \brief Defines the constraint evaluator class for ellipsoids
+*/
+
+// need to declare these class methods with __device__ qualifiers when building in nvcc
+// DEVICE is __host__ __device__ when included in nvcc and blank when included into the host compiler
+#ifdef __HIPCC__
+#define DEVICE __device__
+#else
+#define DEVICE
+#endif
+
+//! Class for evaluating ellipsoid constraints
+/*! <b>General Overview</b>
+    EvaluatorConstraintManifold is a low level computation helper class to aid in evaluating particle constraints on a
+    ellipsoid. Given a ellipsoid at a given position and radii, it will find the nearest point on the Manifold to a given
+    position.
+*/
+class EvaluatorConstraintManifold
+    {
+    public:
+        //! Constructs the constraint evaluator
+        /*! \param _P Position of the ellipsoid
+            \param _rx   Radius of the ellipsoid in the X direction
+            \param _ry   Radius of the ellipsoid in the Y direction
+            \param _rz   Radius of the ellipsoid in the Z direction
+
+            NOTE: For the algorithm to work, we must have _rx >= _rz, ry >= _rz, and _rz > 0.
+        */
+        DEVICE EvaluatorConstraintManifold(Scalar3 _L, Scalar3 _R, manifold_enum::surf _surf)
+            : L(_L), R(_R), surf(_surf)
+            {
+            }
+
+        //! Evaluate the closest point on the ellipsoid. Method from: http://www.geometrictools.com/Documentation/DistancePointEllipseManifold.pdf
+        /*! \param U unconstrained point
+
+            \return Nearest point on the ellipsoid
+        */
+
+        DEVICE Scalar implicit_function(const Scalar3& U)
+        {
+	    Scalar3 delta;
+            switch(surf){
+              case manifold_enum::gyroid:
```

Comment:
This code is a little difficult to maintain/test, but it isn't bad given the limitations of the GPU and given that the individual computations are relatively small.  However, we should unify the GPU and CPU code paths to both use the same code. We could either use the switch/case code and remove the need for the independent Manifold classes, or follow the template pattern used in the rest of HOOMD.

The alternative is to follow the path we do in HPMC and pair potentials: Implement each of the manifolds in its own `ManifoldShape` class (no virtual methods, no base class). Then template the integrators on the type of the Manifold and export separate classes to Python (i.e. TwoStepNVECylinder, TwoStepNVEEllipsoid, TwoStepNVEGyroid, ...)

@b-butler @vyasr  what do you think?

Reply:
If I'm understanding correctly from a skim of the code, right now CPU evaluators are passed instances of `Manifold`, then evaluate the function and its derivatives as needed to perform the iterative evaluation of the Rattle scheme. This code path looks basically identical to what I originally had written. Meanwhile, GPU evaluators use this class `EvaluatorConstraintManifold` that is constructed from the properties of a `Manifold` when the GPU integration method is constructed and contains information about all the known types of manifolds, using this switch statement. This code path is entirely new from what I originally had.

Assuming that I've understood the code correctly, I think switching to header-only, inheritance-free manifolds and templating the integrator is the right approach. The inheritance of manifolds is nice for consistency (and it would be nice if we could do the same for pair potential evaluators), but it's easy enough to document the requirements in the interest of enabling the templating code path. Having all known manifolds embedded into one file's switch statement is messier and less extensible IMO.

Reply:
@joaander @SchoeniPhlippsn is this statement accurate? Ping when you think this PR is in a suitable place for me to review again, and just let me know which version of this I should be expecting.

Reply:
Phillip is rewriting this to use the template evaluators. I think he is almost done.